### PR TITLE
♻️  [RUMF-1178] Improve log assembly part-3

### PR DESCRIPTION
**Pull request by amortemousque**
## Motivation

Improve the Logs assembly.

Since the Logs assembly was intertwined with the rest of the logic, I had a hard time to limit the modifications so
the PR is quite big.  However, the flow is now very similar to RUM, so you should get use to it relatively fast.
Let me know if it is indigestible, I'll try to break it down.

## Changes

- Create a `CommonContext` to encapsulate the current `date`, `view` and `loggerGlobalContext`
- Create a Log `LifeCycle` with `RAW_LOG_COLLECTED` and `LOG_COLLECTED`
- Assemble the different contexts in `startLogsAssembly`
- Add logger  getLevel, getHandler and getContext
- Pass down the main logger to startLogsAssembly to be assembled for all collection

## Testing

- [ ] Local
- [ ] Staging
- [x] Unit
- [ ] End to end

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/main/CONTRIBUTING.md) documentation.

### CODE REVIEW

File: packages/logs/src/domain/logger.ts [hunk 1]
```diff
@@ -64,22 +60,42 @@ export class Logger {
...
+) {
+  const contextManager = createContextManager()
+  // Todo: merge the logger name into the context in logger collection in next major version
+  contextManager.set(assign({}, context, { logger: { name } }))
+
+  return {
+    handlerType,
+    level,
+    name,
+    contextManager,
```
Comment (bcaudan) [inline packages/logs/src/domain/logger.ts:99]: **💬 suggestion:** ‏I think we should handle separately options used to build the current message (name, context) from the one used to "handle" this message (handler, level).

We could partially build the message inside the logger like we did before:
```ts
combine({ message, status }, this.contextManager.get(), messageContext)
```

and let that only override more general contexts (global, rum, service/env/version) during assembly.

wdyt?

File: packages/logs/src/domain/logsCollection/logger/loggerCollection.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+export function startLoggerCollection(lifeCycle: LifeCycle) {
+  function addLog(logsMessage: LogsMessage, loggerOptions: LoggerOptions, savedCommonContext?: CommonContext) {
+    const messageContext = logsMessage.context
+
+    if (isAuthorized(logsMessage.status, HandlerType.console, loggerOptions)) {
+      display.log(
+        `${logsMessage.status}: ${logsMessage.message}`,
+        combine(loggerOptions.contextManager.get(), messageContext)
+      )
+    }
```
[the remark this answers is not in the document]
Comment (amortemousque) [reply packages/logs/src/domain/logsCollection/logger/loggerCollection.ts:24]: I suppose it could always be printed for this type of error :)

[the remark this answers is not in the document]
Comment (bcaudan) [reply packages/logs/src/domain/logsCollection/logger/loggerCollection.ts:24]: 🤔 I had mind that we had some kind of global parameters for handler type but in fact no, it is only logger specific :facep

File: packages/logs/src/boot/logsPublicApi.spec.ts [hunk 3]
```diff
@@ -222,67 +212,24 @@ describe('logs entry', () => {
 
     it('logs a message', () => {
       LOGS.logger.log('message')
-
-      expect(getLoggedMessage(0)).toEqual({
-        context: {
```
Comment (amortemousque) [inline packages/logs/src/boot/logsPublicApi.spec.ts:227]: I removed the common context check since it's only combined assemble stage

File: packages/logs/src/domain/logsCollection/console/consoleCollection.spec.ts [hunk 4]
```diff
@@ -1,120 +1,100 @@
...
     console.error('foo', 'bar')
 
-    expect(sendLogSpy.calls.mostRecent().args[0].error).toEqual({
+    expect(rawLogs[0].rawLog.error).toEqual({
       origin: ErrorSource.CONSOLE,
       stack: undefined,
     })
   })
-
-  it('should not print the log twice when console handler is enabled', () => {
```
Comment (amortemousque) [inline packages/logs/src/domain/logsCollection/console/consoleCollection.spec.ts:103]: This test now requires that `consoleCollection` and `loggerCollection` start so I moved to startLogs.spec

File: packages/logs/src/domain/logger.spec.ts [hunk 5]
```diff
@@ -30,126 +30,83 @@ describe('Logger', () => {
       expect(getLoggedMessage(0).status).toEqual(StatusType.info)
     })
 
-    it("'logger.log' should set 'logger' origin when ff forward-logs enabled", () => {
```
Comment (amortemousque) [inline packages/logs/src/domain/logger.spec.ts:33]: Moved to loggerCollection.spec

File: packages/logs/src/domain/logsCollection/console/consoleCollection.ts [hunk 6]
```diff
@@ -19,24 +19,25 @@ const LogStatusForApi = {
...
-          origin: ErrorSource.CONSOLE, // Todo: Remove in the next major release
-          stack: log.stack,
-        },
-      }
-    }
-    if (isExperimentalFeatureEnabled('forward-logs')) {
-      messageContext.origin = ErrorSource.CONSOLE
-    }
-    sender.sendToHttp(log.message, messageContext, LogStatusForApi[log.api])
+    lifeCycle.notify(LifeCycleEventType.RAW_LOG_COLLECTED, {
```
Comment (bcaudan) [inline packages/logs/src/domain/logsCollection/console/consoleCollection.ts:27]: **🥜 nitpick:** ‏we could inline this function to avoid to name it

File: packages/logs/src/domain/logsCollection/console/consoleCollection.ts [hunk 7]
```diff
@@ -19,24 +19,25 @@ const LogStatusForApi = {
   [ConsoleApiName.warn]: StatusType.warn,
   [ConsoleApiName.error]: StatusType.error,
 }
-export function startConsoleCollection(configuration: LogsConfiguration, sender: Sender) {
+export function startConsoleCollection(configuration: LogsConfiguration, lifeCycle: LifeCycle) {
   const consoleObservable = initConsoleObservable(configuration.forwardConsoleLogs)
   const consoleSubscription = consoleObservable.subscribe(reportConsoleLog)
```
Comment (bcaudan) [inline packages/logs/src/domain/logsCollection/console/consoleCollection.ts:24]: **🥜 nitpick:** ‏`consoleObservable` could be inlined

File: packages/logs/src/domain/logsCollection/logger/loggerCollection.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+    if (isAuthorized(logsMessage.status, HandlerType.console, logger)) {
+      display.log(`${logsMessage.status}: ${logsMessage.message}`, combine(logger.getContext(), messageContext))
+    }
+
+    lifeCycle.notify(LifeCycleEventType.RAW_LOG_COLLECTED, {
+      rawLog: {
+        message: logsMessage.message,
+        status: logsMessage.status,
+        origin: isExperimentalFeatureEnabled('forward-logs') ? ErrorSource.LOGGER : undefined,
```
Comment (bcaudan) [inline packages/logs/src/domain/logsCollection/logger/loggerCollection.ts:27]: **💬 suggestion:** ‏related to previous PRs, should we rename `ErrorSource` since it does not seem to be related to error anymore?

File: packages/logs/src/domain/logsCollection/report/reportCollection.ts [hunk 9]
```diff
@@ -18,18 +18,16 @@ const LogStatusForReport = {
...
 
   function logReport(report: RawReport) {
     let message = report.message
-    const messageContext: Partial<LogsEvent> = {
-      origin: ErrorSource.REPORT,
-    }
-    const logStatus = LogStatusForReport[report.type]
-    if (logStatus === StatusType.error) {
-      messageContext.error = {
+    const status = LogStatusForReport[report.type]
```
Comment (bcaudan) [inline packages/logs/src/domain/logsCollection/report/reportCollection.ts:27]: **🥜 nitpick:** we could inline this function to avoid to name it‏

File: packages/logs/src/rawLogsEvent.types.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
+import type { Context, ErrorSource, TimeStamp } from '@datadog/browser-core'
+import type { StatusType } from './domain/logger'
+
+export interface RawLogsEvent {
```
Comment (bcaudan) [inline packages/logs/src/rawLogsEvent.types.ts:4]: **💬 suggestion:** ‏could it be interesting to document the different raw types, console, report, logger, ...?

File: packages/logs/src/domain/logsCollection/logger/loggerCollection.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  [StatusType.error]: 3,
+}
+
+export function startLoggerCollection(lifeCycle: LifeCycle) {
+  function handleLog(logsMessage: LogsMessage, logger: Logger, savedCommonContext?: CommonContext) {
+    const messageContext = logsMessage.context
+
+    if (isAuthorized(logsMessage.status, HandlerType.console, logger)) {
+      display.log(`${logsMessage.status}: ${logsMessage.message}`, combine(logger.getContext(), messageContext))
+    }
```
Comment (BenoitZugmeyer) [inline packages/logs/src/domain/logsCollection/logger/loggerCollection.ts:21]: **💭 thought:** ‏at first glance I expected that the `display.log` call should be done in another module using `lifeCycle.subscribe(LOG_COLLECTED, ...)`. But:

* it seems that LOG_COLLECTED (and the whole LifeCycle, and the assembly module) only expects to handle logs that should be sent to datadog. It's a bit unexpected but seems reasonable.

* I think that we should keep the `display.log` call close to the actual `Logger.*` usage, i.e. not moving it in a LifeCycle `subscribe` callback, but maybe moving it back to the `Logger` class. With #1501 we'll have a stack trace displayed below "error" and "warn" logs, which is great! But having `display.log` deep into the stack trace adds a bit of noise and impacts the developper experience. For example, with your PR, the stack trace would look like this:
Screenshot 2022-04-22 at 17 56 38
Which is not dramatic, but we might want to minimize this stack trace :)

**Comment by BenoitZugmeyer** [comment]
I'm actually curious about your decision to remove the `LoggerOptions` and extend the `Logger` class instead :) I thought it was a bit nicer to have a "dumb" state used accross the SDK instead of a `Logger` instance that:
* constrains us to the current design, as it is also publicly exposed
* should not be used internally by convention (except to access their state)

File: packages/logs/src/domain/assembly.ts [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+      configuration.eventRateLimiterThreshold,
+      reportAgentError
+    ),
+    [StatusType.debug]: createEventRateLimiter(
+      StatusType.debug,
+      configuration.eventRateLimiterThreshold,
+      reportAgentError
+    ),
+    ['custom']: createEventRateLimiter('custom', configuration.eventRateLimiterThreshold, reportAgentError),
+  }
```
[the remark this answers is not in the document]
Comment (amortemousque) [reply packages/logs/src/domain/assembly.ts]: 👍

File: packages/logs/src/domain/lifeCycle.ts [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    return {
+      unsubscribe: () => {
+        this.callbacks[eventType] = this.callbacks[eventType]!.filter((other) => callback !== other)
+      },
+    }
+  }
+}
+
+export interface RawLogCollectedData<E extends RawLogsEvent = RawLogsEvent> {
+  rawLog: E
```
[the remark this answers is not in the document]
Comment (amortemousque) [reply packages/logs/src/domain/lifeCycle.ts]: Indeed

File: packages/logs/src/boot/logsPublicApi.ts [hunk 14]
```diff
@@ -104,19 +115,7 @@ export function makeLogsPublicApi(startLogsImpl: StartLogs) {
...
-          view: {
-            referrer: document.referrer,
-            url: window.location.href,
-          },
-        },
-        globalContextManager.get()
-      )
-    )
+  function createLogger(options: LoggerOptions) {
+    return new Logger(options, (...params) => addLogStrategy(...params))
```
[the remark this answers is not in the document]
Comment (amortemousque) [reply packages/logs/src/boot/logsPublicApi.ts]: Sure

File: packages/logs/src/domain/assembly.ts [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        logger.getContext(),
+        messageContext
+      )
+
+      if (
+        configuration.beforeSend?.(log) === false ||
+        (log.error?.origin !== ErrorSource.AGENT &&
+          (logRateLimiters[log.status] ?? logRateLimiters['custom']).isLimitReached()) ||
+        // Todo: [RUMF-1230] Move this check to the logger collection in the next major release
+        !isAuthorized(rawLog.status, HandlerType.http, logger)
```
[the remark this answers is not in the document]
Comment (amortemousque) [reply packages/logs/src/domain/assembly.ts]: true

File: packages/logs/src/domain/assembly.ts [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+      const commonContext = savedCommonContext || getCommonContext()
+      const log = combine(
+        { service: configuration.service, session_id: session.id },
+        { date: commonContext.date, view: commonContext.view },
+        commonContext.context,
+        getRUMInternalContext(startTime),
+        rawLog,
+        logger.getContext(),
+        messageContext
+      )
```
[the remark this answers is not in the document]
Comment (amortemousque) [reply packages/logs/src/domain/assembly.ts]: It would mean that at assembly level, we can not guarantee the format for `rawLogsEvent` coming from `Loggers` since it can be overridden by the user. This can be problematic for logic relying on certain field like `date`, `origin` or `status`. Wdyt?

File: packages/logs/src/domain/logsCollection/logger/loggerCollection.ts (hunk 8 above)
Comment (amortemousque) [reply packages/logs/src/domain/logsCollection/logger/loggerCollection.ts]: Indeed. I'm wondering if the issue is only on naming. Now that it serve `log origin` and `rum error source`, I feel like it starts to be different concepts. We already have divergence:
```
  AGENT: 'agent',
  CONSOLE: 'console',
  CUSTOM: 'custom', // specific RUM
  LOGGER: 'logger', // specific Logs
  NETWORK: 'network', // specific Logs
  SOURCE: 'source',
  REPORT: 'report',
```
IMO, there is two options:

- keep a single list in coe packahe with a better name like: `apiSource` 

- have two different list:

`logOrigins` in Logs package
`errorSource` in RUM core package

Wdyt?

File: packages/logs/src/rawLogsEvent.types.ts (hunk 10 above)
Comment (amortemousque) [reply packages/logs/src/rawLogsEvent.types.ts]: Indeed it could be nice to have.

File: packages/logs/src/domain/logsCollection/logger/loggerCollection.ts (hunk 11 above)
Comment (amortemousque) [reply packages/logs/src/domain/logsCollection/logger/loggerCollection.ts:21]: Yes right. Since this PR is quite big and it takes times to validate, I'm in favor of working on this improvement while doing https://github.com/DataDog/browser-sdk/pull/1501

File: packages/logs/src/domain/assembly.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (bcaudan) [reply packages/logs/src/domain/assembly.ts]: FMU, those fields can still be overwritten by the customer. I would find it less surprising to use customer value as soon as possible rather than using our value for an internal behavior (like associate RUM context from the date) and then overwrite our value by the customer one which would make thing quite hard to debug.

File: packages/logs/src/domain/logsCollection/logger/loggerCollection.ts (hunk 8 above)
Comment (bcaudan) [reply packages/logs/src/domain/logsCollection/logger/loggerCollection.ts]: lgtm, we should probably postpone that to a next PR though

Comment (amortemousque) [reply packages/logs/src/domain/logsCollection/logger/loggerCollection.ts]: What option exactly look good to you? :)

Comment (bcaudan) [reply packages/logs/src/domain/logsCollection/logger/loggerCollection.ts]: 😅 two different lists

File: packages/logs/src/boot/logsPublicApi.ts [hunk 17]
```diff
@@ -91,13 +91,7 @@ export function makeLogsPublicApi(startLogsImpl: StartLogs) {
...
 
     createLogger: monitor((name: string, conf: LoggerConfiguration = {}) => {
-      customLoggers[name] = new Logger(
-        (...params) => handleLogStrategy(...params),
-        name,
-        conf.handler,
-        conf.level,
-        conf.context
-      )
+      customLoggers[name] = new Logger(handleLogStrategy, name, conf.handler, conf.level, conf.context)
```
Comment (bcaudan) [inline packages/logs/src/boot/logsPublicApi.ts:94]: **❗ warning:** FMU using `handleLogStrategy` directly here should break loggers created before the init 🤔 
do we have a test for this one?‏

File: packages/logs/src/rawLogsEvent.types.ts [hunk 18]
```diff
@@ -1,29 +1,67 @@
...
-  date?: TimeStamp
+export type RawLogsEvent =
+  | RawConsoleLogsEvent
+  | RawNetworkLogsEvent
+  | RawLoggerLogsEvent
+  | RawAgentLogsEvent
+  | RawReportLogsEvent
+  | RawRuntimeLogsEvent
+
+export type RawLogsEventOrigin = RawLogsEvent['origin']
```
Comment (bcaudan) [inline packages/logs/src/rawLogsEvent.types.ts:12]: **❓ question:** ‏is it used somewhere?

File: packages/logs/src/rawLogsEvent.types.ts [hunk 19]
```diff
@@ -1,29 +1,67 @@
...
-    [k: string]: unknown
-  }
-  http?: {
+  error?: Error
+}
+
+interface CommonRawErrorLogsEvent extends CommonRawLogsEvent {
+  status: typeof StatusType.error
+  error: Error
+}
```
Comment (bcaudan) [inline packages/logs/src/rawLogsEvent.types.ts:30]: **🥜 nitpick:** ‏personal preference, I would inline this interface to avoid to navigating across the hierarchy to understand what we actually have in an interface

File: packages/logs/src/rawLogsEvent.types.ts [hunk 20]
```diff
@@ -1,29 +1,67 @@
...
+
+export interface RawConsoleLogsEvent extends CommonRawLogsEvent {
+  origin: typeof ErrorSource.CONSOLE
+}
+
+export interface RawLoggerLogsEvent extends CommonRawLogsEvent {
+  origin: typeof ErrorSource.LOGGER
+}
+
```
Comment (bcaudan) [inline packages/logs/src/rawLogsEvent.types.ts:39]: **💭 thought:** ‏I would have love to have `logger?: {name: string}` here

File: packages/logs/src/rawLogsEvent.types.ts [hunk 21]
```diff
@@ -1,29 +1,67 @@
...
-    [k: string]: unknown
-  }
-  error?: {
-    kind?: string
-    origin: ErrorSource
-    stack?: string
-    [k: string]: unknown
-  }
-  http?: {
+  error?: Error
```
Comment (bcaudan) [inline packages/logs/src/rawLogsEvent.types.ts:24]: **💭 thought:** ‏it could be nice to have a design allowing to have `date: TimeStamp` here

File: packages/logs/src/domain/assembly.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (amortemousque) [reply packages/logs/src/domain/assembly.ts]: Fair enough. Since it's a non negligible change I'm in favor of doing it when tackling the [RUMF-1230](https://datadoghq.atlassian.net/browse/RUMF-1230)
Wdyt?

File: packages/logs/src/domain/logsCollection/logger/loggerCollection.ts (hunk 8 above)
Comment (amortemousque) [reply packages/logs/src/domain/logsCollection/logger/loggerCollection.ts]: Yes I'll do it in a following PR

File: packages/logs/src/rawLogsEvent.types.ts (hunk 20 above)
Comment (amortemousque) [reply packages/logs/src/rawLogsEvent.types.ts:39]: Yes me too. It directly related to https://github.com/DataDog/browser-sdk/pull/1497#discussion_r853177027
and It will be done with [RUMF-1230](https://datadoghq.atlassian.net/browse/RUMF-1230)

File: packages/logs/src/boot/logsPublicApi.ts (hunk 17 above)
Comment (amortemousque) [reply packages/logs/src/boot/logsPublicApi.ts]: Indeed. I fixed it and added a test.

File: packages/logs/src/rawLogsEvent.types.ts [hunk 22]
```diff
@@ -37,9 +31,11 @@ export interface RawLoggerLogsEvent extends CommonRawLogsEvent {
   origin: typeof ErrorSource.LOGGER
 }
 
-export interface RawNetworkLogsEvent extends CommonRawErrorLogsEvent {
+export interface RawNetworkLogsEvent extends CommonRawLogsEvent {
   date: TimeStamp
```
Comment (bcaudan) [inline packages/logs/src/rawLogsEvent.types.ts:35]: **🥜 nitpick:** ‏date should not be needed here anymore

**Review by bcaudan — APPROVED** [review]
